### ISSUE TITLE
[Optimizer] Stateful L1-spill validation runs before trace-hoist; per-trace-graph allocation unmodeled → runtime L1 CB-vs-buffer clash (OOM)

### ISSUE THREAD

**Issue by rpavlovicTT** [tenstorrent/tt-mlir#9123]
## Summary

The stateful L1-spill validation (`GreedyL1SpillManagement`, run by `createTTNNPipelineAnalysisPasses`) executes on the **contiguous `main` function**, *before* `createTTNNTraceHoistTransform` splits `main` into per-trace functions (`trace_0_main` … `trace_N_main`). Because L1 is validated against the un-split graph, the compile-time memory view diverges from the runtime's per-trace-graph allocation, and the pass can **miss a real runtime L1 clash / OOM**.

Observed as a hard failure: `vllm_qwen2_5_coder_32b_instruct_qb2_tp` on QB2 throws

```
TT_THROW: Statically allocated circular buffers in program <id> clash with L1 buffers on core range [0-0 - 0-0].
L1 buffer allocated at 447488 and static circular buffer region ends at 471808
```

byte-reproducible across hosts.

## Root cause (evidence)

The allocation **algorithm is the same** at compile and runtime — compile-time placement uses the real allocator via `extract_mock_allocator_state(...).with_allocations(...)` (`lib/OpModel/TTNN/TTNNOpModel.cpp`), which pins live records and real-allocates. The divergence is entirely in the **inputs (allocation order + live-set)**, and its source is pipeline ordering:

- `createTTNNPipelineAnalysisPasses` → **GreedyL1SpillManagement** runs first, on the contiguous `main`.
- `createTTNNTraceHoistTransform` runs *later* and **splits `main` into separate per-trace programs**, each with its own allocation lifecycle at runtime.

So at **compile** time the vocab-sized logits buffer (`32×152064×f32`, ~19 MB) is allocated **once**, high (e.g. 1327104), and pinned — it is never placed low, so no clash is seen. At **runtime** the logits is re-allocated **per trace graph**; the sampling graph (`trace_5_main`, holding `topk` + a vocab-sized `ttnn.slice_static`) churns dozens of vocab-sized transients (up to ~84 MB), which shapes the free-list so a logits copy lands **low at 447488 — inside the slice's static CB region `[111360, 471808)`** → clash. Each trace graph starts near-clean (only a few persistent scalars pinned at the top of L1), so the clash is driven by the within-graph churn that the pre-hoist validation never models.

Secondary gap: even within its own address space, the compile-time fit query only checks **CB headroom** (does `cbPeakUsage` of free space exist), not the fixed-address `lowest_occupied < cb_region_end` collision that runtime's `validate_circular_buffer_region` enforces.

## Fix direction

Run the stateful L1-spill (and a fixed-address CB-collision check mirroring `validate_circular_buffer_region`) **after `TTNNTraceHoistTransform`, per trace function**, seeding the persistent cross-trace residency, so the compile-time memory view matches the runtime's per-trace-graph reality. This is tensor-agnostic (no per-tensor heuristics) — it makes the two allocation timelines identical.

A headroom-reservation ("static-CB headroom floor") approach was tried and **rejected/reverted**: it masks the symptom, over-reserves L1 (occupancy loss + extra ops), and does not address the ordering root cause.

## Reproduction

2-layer `qwen2.5-coder-32b-instruct` TP on QB2 with **device sampling**:

```
TT_BENCHMARK_NUM_HIDDEN_LAYERS=2 TT_BENCHMARK_DEVICE_SAMPLING=1 \
  pytest -svv "tests/benchmark/test_vllm_benchmarks.py::test_vllm_tp_benchmark[qwen2.5-coder-32b-instruct-qb2-tp]"
```

Note: the benchmark default `cpu_sampling=True` (at `optimization_level>=1`) **masks** the clash — it removes the device sampling graph (no `topk` / vocab-sized `slice_static`). Device sampling must be forced to reproduce.
